### PR TITLE
Tag values are expected to be strings

### PR DESCRIPTION
**Pull request by mparkhe**
- Fixing context providers to ensure that string values are passed for "source type". Since `SourceType` is an enum, converting to/from string involves calling class methods to preserver backward compatibility of types displayed in UI and ensuring that `RestStore` can hydrate protobufs appropriately. Fixed tests to reflect this change.

- Also making tag resolution order and precedence explicit over multiple context providers.

### CODE REVIEW

File: mlflow/store/rest_store.py [hunk 1]
```diff
@@ -152,6 +152,7 @@ def create_run(self, experiment_id, user_id, run_name, source_type, source_name,
         response_proto = self._call_endpoint(CreateRun, req_body)
         run = Run.from_proto(response_proto.run)
         if run_name:
+            # TODO: optimization: This is making 2 calls to backend store. Include with above call.
```
Comment (mparkhe) [inline mlflow/store/rest_store.py]: We can add this to `user_specified_tags` and will `ContextProviders` will auto resolve this as run name. This is a bit of a involved change and not related to this fix. Will handle this separately.

File: tests/store/test_rest_store.py [hunk 2]
```diff
@@ -87,6 +97,27 @@ def test_requestor(self, request):
         creds = MlflowHostCreds('https://hello')
         store = RestStore(lambda: creds)
 
+        user_name = "mock user"
+        run_name = "rest run"
+        source_name = "rest test"
+
+        with mock.patch('mlflow.store.rest_store.http_request_safe') as mock_http, \
```
Comment (mparkhe) [inline tests/store/test_rest_store.py]: added a new test that can catch issues related to `CreateRun` protobuf handling

File: mlflow/tracking/fluent.py [hunk 3]
```diff
@@ -127,21 +127,23 @@ def start_run(run_uuid=None, experiment_id=None, source_name=None, source_versio
...
             user_specified_tags[MLFLOW_PROJECT_ENTRY_POINT] = entry_point_name
 
         tags = context.resolve_tags(user_specified_tags)
 
+        # Polling resolved tags for run meta data : source_name, source_version,
+        # entry_point_name, and source_type which is store in RunInfo for backward compatibility.
         active_run_obj = MlflowClient().create_run(
             experiment_id=exp_id_for_run,
             run_name=run_name,
-            source_name=tags.get(MLFLOW_SOURCE_NAME),  # For backwards compatability
```
Comment (smurching) [inline mlflow/tracking/fluent.py]: Micronit: I actually kinda like the "For backwards compatibility" comments b/c they help clarify what we might be able to remove with MLflow 1.0, but ok w/ deleting them too

**Review by smurching — APPROVED** [review]
LGTM thanks for the fix + new test!

File: mlflow/tracking/fluent.py (hunk 3 above)
Comment (mparkhe) [reply mlflow/tracking/fluent.py]: You are right! Done ... added smaller annotation to avoid text wrapping.

**Pull request merged**